SSMProblems/src/utils/forward_simulation.jl

```
@@ -8,25 +10,15 @@ export sample
 
 Simulate a trajectory of length `T` from the state space model.
 
-Returns a tuple `(x0, xs, ys)` where `x0` is the initial state, `xs` is a vector of latent states,
-and `ys` is a vector of observations.
+Returns a tuple `(xs, ys)` where `xs` is a vector of latent states (including the initial
+state) and `ys` is a vector of observations.
 """
-function sample(
-    rng::AbstractRNG, model::StateSpaceModel{<:Real,LD,OP}, T::Integer; kwargs...
-) where {LD,OP}
-    T_dyn = eltype(LD)
-    T_obs = eltype(OP)
-
-    xs = Vector{T_dyn}(undef, T)
-    ys = Vector{T_obs}(undef, T)
-
-    x0 = simulate(rng, model.dyn; kwargs...)
+function sample(rng::AbstractRNG, model::StateSpaceModel, T::Integer; kwargs...)
+    xs = OffsetVector(fill(simulate(rng, model.prior; kwargs...), T + 1), -1)
```

This doesn't seem compatible for the philosophy of this PR (i.e. that the prior and dynamics could return different types). Since we create a container from the prior sample, we're assuming the rest will have the same type.

Two solutions:
- We remove the OffsetVector: I quite liked this addition, but if it helps keep the prior/dynamics separate that's a good enough motivation to ditch it for me
- We let this fail, potentially adding a try-catch to tell the user, it's likely because their prior/dynamics types don't match (they can always cast their prior output to the expected type, e.g. `Dual`).

---

This is actually a really stellar point.

I quite enjoy the use of OffsetVectors, although I cannot help but feel the same way. I think we should revert to the previous behavior and if we nail the type promotion, we can patch it in later.